src/mpi/init/init_async.c

```diff
@@ -91,6 +91,18 @@ static void progress_fn(void *data)
     MPID_Thread_cond_signal(&progress_cond, &mpi_errno);
     MPIR_Assert(!mpi_errno);
 
+    /* Tell the main thread that it's ok to destroy everything */
+    MPID_Thread_mutex_lock(&progress_mutex, &mpi_errno);
+    MPIR_Assert(!mpi_errno);
+
+    progress_thread_done = 2;
```

Comment:
what is 2?

Reply:
It needs a second synchronization. `2` is just an arbitrary status, just like `1` is for first stage synchronization. I guess it is better to define a set of symbols. Would you like me to do that?

Reply:
I don't understand why we have to signal the main thread twice during finalize. Why is this required when it wasn't before?

Reply:
I believe the problem always existed but the behavior is kernel or os dependent. For example, I couldn't reproduce it on my workstation, but it manifests on jenkins.

The first stage synchronizes on when progress thread finishes MPI tasks. But after that, if the main thread somehow detects the `progress_thread_done` is 1 before before entering the while loop, it won't even wait on the condition variable -- this happens quite often. Then it is a race condition between the main thread destroying the condition variable and the progress thread signaling the condition variable. I don't quite understand why signaling an already destroyed condition variable hangs on our jenkins server, but the only way to prevent that is to have a second synchronization.

Reply:
Also, the main thread can't simply  wait on the condition variable without checking `progress_thread_done` first either, because there is a chance progress thread went ahead signals the condition before the main thread wait and leave it hanging, so the first synchronization is also necessary.

Reply:
It's troubling that we do not understand how to synchronize with this thread.

Reply:
https://github.com/pmodels/mpich/blob/d1665bb7594358251d82d44952ecd6afd4993092/src/mpi/init/init_async.c#L22-L33

Reply:
One thing I noticed is that we do not join the progress thread, and never have. That is surprising to me. Are we leaving zombies? Is that what is causing the Jenkins nodes to eventually fall over?

Reply:
https://stackoverflow.com/questions/6594388/why-we-should-use-join-in-threads , top answer.

Reply:
When I tested on jenkins, the progress thread hangs at `MPID_Thread_cond_signal`.

Reply:
> https://stackoverflow.com/questions/6594388/why-we-should-use-join-in-threads , top answer.

Huh?

Reply:
The `join` is simply the main thread waiting for other threads to exit. It is a synchronization. A thread is not a process, so I don't think it can become a zombie -- the process as a whole will not exit. I guess we could use `thread_join` here for the second synchronization -- which I believe is a busy loop anyway. Unfortunately, we don't have MPL macros for join.

Reply:
Now that you mentioned, join is what we needed here (in stead of a manual second synchronization). We'll need add MPL_thread_join for every thread packages.